Additions Parachains Host Implementers Guide (#4200)
w3f/polkadot-wiki: docs/learn/learn-validator.md

{{ kusama: <RPC network="kusama" path="query.staking.validatorCount" defaultValue={1000}/> :kusama }}
validators randomly selected (per epoch) to participate in the validation, creating a validator pool
of 200 para-validators.

Validators perform two main functions:

1. **Verifying** that the information contained in an assigned set of parachain blocks is valid.
   They receive parachain block candidates from the [collators](./learn-collator.md) together with a
   Proof-of-Validity (PoV). The para-validators then check if the block candidates are valid.
   Candidates that gather enough signed validity statements are considered _backable_.
2. **Participating** in the consensus mechanism to produce the relay chain blocks based on validity
   statements from other validators. These validators are called block authors, they are selected by
   [BABE](./learn-consensus.md/#block-production-babe) and can note up to one backable candidate for
   each parachain to include in the relay chain. A backable candidate included in the relay chain is
   considered _backed_ in that fork of the chain.

> Validators also participate to the so-called **availability distribution**. In fact, once the

Reviewer: ```suggestion
Validators also contribute to the so-called **availability distribution**. In fact, once the
```

candidate is backed in a fork of the relay chain, it is still _pending availability_, i.e. it is not
included as part of the parachain until it is proven avaialable (together with the PoV). Information
regarding the availability of the candidate will be noted in the following relay chain blocks. Only
when there is enough information, the candidate is considered a full parachain block or _parablock_.

> Validators also participate to the so-called [**approval process**](#approval-process). Once the

Reviewer: ```suggestion
Validators also participate in the the so-called [**approval process**](#approval-process). Once the
```

parablock is considered available and part of the parachain, it is still _pending approval_. Because

> para-validators are a small subset of all validators, there a risk that by chance the majority of

Reviewer: ```suggestion
para-validators are a small subset of all validators, there is a risk that by chance the majority of
```

for their activities.

## Approval Process

Having a bad parablock on the relay chain is not catastrophic as long as the block is not approved
and finalized by the finality gadget [GRANDPA](./learn-consensus.md/#finality-gadget-grandpa). If
the block is not finalized, the fork on the chain containing that block can be ignored in favor of
another fork containing good blocks. Dealing with a bad parablock includes the following stages:

- Detection: the bad block must be detected by honest validators.
- Escalation: the honest validators must send that block for checks to all validators. A dispute
  starts.
- Consequences: the chain is reverted and all malicious validators are slashed.

The result of the dispute must be transplantable to all other forks so that malicious validators are
slashed in all possible histories and so that honest validators will ignore any forks containing
that parablock.

:::info Parablocks vs Relay-Chain blocks

It is important to understand that a relay chain block contains many parablocks. Thus, it makes more
sense to think of relay-chain blocks as having been approvead instead of parablocks have been

> approved. A rely-chain block containing a bad parablock must be reverted, while a relay-chain block

Reviewer: ```suggestion
sense to think of relay-chain blocks as having been approvead instead of parablocks that have been
approved. A relay-chain block containing a bad parablock must be reverted, while a relay-chain block
```

containing only approved parablocks can be considered approved as long as its parent relay-chain
block is also approved. Thus, the validity of a relay-chain block depends on the validity of its
ancestry.

:::

The Approval Process is divided into two parts:

- **[Assignments][]** determine which validators perform approval checks on which candidates,
  ensuring each candidate receives enough random checkers. This stage tracks approval votes to
  identify when [no-show][] approval checks take suspiciously long. It also tracks relay chain
  [equivocations](../maintain/maintain-guides-best-practices-to-avoid-slashes.md/#equivocation) to
  determine when adversaries possibly gained foreknowledge about assignments and adding more checks
  in those cases. Assignees determine their own assignments to check specific candidates using two
  or three [assignment criteria][], which are based upon two possible [stories][] about the relay
  chain block that included the candidate (i.e. declared the candidate available). [Assignment

>   notices][] are gossiped among nodes so that all validators know which validators should check

Reviewer: If a link in this format is forced into a newline by prettier (due to spaces in the naming) does it cause the link to break?  From the syntax highlighting here I would this `Assignment notices` link is broken.

  which candidates, and if any candidate requires more checkers.
- **Approval checks** performs the checks by obtaining the candidate, verify its validity, and
  sending out the approval vote or initiating a dispute. Approval checks have a no-show timeout
  window (i.e. longer than one relay chain slot) to succeed in reconstructing the candidate block,
  redo its erasure coding to check the candidate receipt, and recheck the candidate block itself. A
  validator becomes tagged as no-show if does not approve or dispute within the no-show timeout
  window. Because validators can be overloaded with assignments, they can intentionally delay
  sending their assignment notice to avoid creating no-shows (see more in [Assignment

>   postponement][]).

Reviewer: same as above here

a) assignments notices and approval votes are recorded in a relay chain block, and b) in another
relay chain block notes are fed into the approval code.

The gossiped messages are of two types, assignment notices and approval votes, and are singed with
[approval keys][]. Such keys are part of the [session keys](./learn-cryptography.md/#session-keys)
used by validators. Briefly, approval keys are:

- **Approval assignment keys** that are sr25519 keys used only for assignment criteria
  [VRF](./learn-randomness.md/#vrf).
- **Approval vote keys** that are ed25519 and would only sign off on a candidate parablock validity.

:::info

For detailed information about the approval process see dedicated section in
[The Polkadot Parachain Host Implementers' Guide](https://paritytech.github.io/polkadot/book/protocol-approval.html).

:::

Accepting a parablock is the end result of having passed through the detection stage without
dispute, or having passed through and escalation/dispute stage with a positive outcome.

## Disputes

All parachain blocks that are in the finalized relay chain should be valid. This, does not apply to

> backed blocks that are not included. To ensure nothing invalid end up in the finalized relay chain

Reviewer: ```suggestion
backed blocks that are not included. To ensure nothing invalid ends up in the finalized relay chain
```

there are approval checks (described above) and disputes. The latter ensures that each attempt to
include something invalid is caught and the offending validators are punished.

Disputes are _independent from a particular fork_, while backing and approval operate on particular
forks. The approval voting stops if an alternative fork (which might not contain the
currently-approved candidate) is finalized. In fact, the sole purpose of the approval process is to
make sure invalid blocks are not finalized. However, even though the danger is past and the
offending validators did not manage to get the invalid block approved, those validators need to get
slashed for the attempt.

A dispute stems from a disagreement between teo or more validators. For this to happen a bad actor

> needs to distribute and invalid block to honest validators. Scenarios leading to a dispute can be

Reviewer: ```suggestion
A dispute stems from a disagreement between two or more validators. For this to happen, a bad actor
needs to distribute an invalid block to honest validators. Scenarios leading to a dispute can be
```
